### PR TITLE
Add search module. Add sorting

### PR DESCRIPTION
**Pull request by MAKARD**

### CODE REVIEW

File: receipe-app-api/src/categories/schemas/category.schema.ts [hunk 1]
```diff
@@ -1,7 +1,15 @@
 import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
 import { ApiProperty } from '@nestjs/swagger';
 
-@Schema()
+@Schema({
+    toJSON: {
+        virtuals: true,
```
Comment (valeskus) [inline receipe-app-api/src/categories/schemas/category.schema.ts:8]: what's virtuals?

File: receipe-app-api/src/search/search.service.ts [hunk 2]
```diff
@@ -1,177 +1,91 @@
...
-        recipes: [],
-        filters: [],
-        sortOptions: [],
-      };
-    }
+    private getSortPipelineBySortOption(sortOption: SortOptions): PipelineStage.Sort['$sort'] {
+        switch (sortOption) {
+            case SortOptions.RELEVANCE: {
+                return {
+                    score: { $meta: "textScore" }
```
Comment (valeskus) [inline receipe-app-api/src/search/search.service.ts]: what're we doing here?

File: receipe-app-api/src/search/search.service.ts [hunk 3]
```diff
@@ -1,177 +1,91 @@
...
-          time: '1:50',
-        },
-      ],
-      filters,
-      sortOptions,
-    };
-  }
+    async search(params: SearchDto): Promise<SearchResultsDto> {
+        const mainAggregation = this.recipeModel.aggregate([
+            { $match: { $text: { $search: params.search } } },
```
Comment (valeskus) [inline receipe-app-api/src/search/search.service.ts]: what're we doing here?

**Review by valeskus — CHANGES_REQUESTED** [review]

File: receipe-app-api/src/search/search.service.ts [hunk 4]
```diff
@@ -1,177 +1,43 @@
...
+            filters: new FiltersDto({
+                inputFilters: params,
+                calories,
+                totalTime,
+                difficulty,
+                mealType,
+                dietType,
+            }),
+            recipes,
+            total: totals[0]?.total || 0,
```
Comment (valeskus) [inline receipe-app-api/src/search/search.service.ts:37]: what're you doing here?

File: receipe-app-api/src/main.ts [hunk 5]
```diff
@@ -9,7 +9,9 @@ import { AppModule } from './app.module';
 
 async function bootstrap() {
   const app = await NestFactory.create(AppModule);
-  app.useGlobalPipes(new ValidationPipe());
+  app.useGlobalPipes(new ValidationPipe({
```
Comment (valeskus) [inline receipe-app-api/src/main.ts:12]: what is the useGlobalPipes?

File: receipe-app-api/src/search/aggregations/categories.aggregation.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { FilterQuery, PipelineStage } from 'mongoose';
+
+interface CategoriesAggregationParams {
```
Comment (valeskus) [inline receipe-app-api/src/search/aggregations/categories.aggregation.ts:5]: what're you doing here?

File: receipe-app-api/src/search/aggregations/ranged-facet.aggregation.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import { FilterQuery, PipelineStage } from 'mongoose';
+
+export interface RangedFacet {
+    operator: string;
+    value: number;
+}
+
+interface RangedFacetAggregationParams<T> {
```
Comment (valeskus) [inline receipe-app-api/src/search/aggregations/ranged-facet.aggregation.ts:8]: what's  ranged?

File: receipe-app-api/src/search/dto/filters.dto.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+const multipleFilterTransformation = <T>(list: Facet<T>, appliedFilter?: Array<T>) => {
+    return filterOutEmpty(list).map((item) => ({
+            ...item,
+            isActive: appliedFilter?.includes(item.value) || false
+        }));
+};
+
+class Filters {
+    constructor(params: FiltersData) {
+        this.calories = singleFilterTransformation(params.calories, params.inputFilters.calories);
```
Comment (valeskus) [inline receipe-app-api/src/search/dto/filters.dto.ts:37]: what's the singleFilterTransformation?

**Review by valeskus — CHANGES_REQUESTED** [review]

File: receipe-app-api/src/search/aggregations/categories.aggregation.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public filter: FilterQuery<unknown> = {};
+
+    public get facet(): Record<string, Array<PipelineStage.FacetPipelineStage>> {
+        return {
+            [this.facetName]: [
+                { $unwind: '$categories' },
+                { $sortByCount: '$categories' },
+                { $match: { '_id.type': 'meal' } },
```
Comment (MAKARD) [inline receipe-app-api/src/search/aggregations/categories.aggregation.ts]: TODO

**Review by valeskus — APPROVED** [review]

**Pull request merged**